pkg/service/egress.go

```diff
@@ -300,68 +351,130 @@ func (s *EgressService) StopEgress(ctx context.Context, req *livekit.StopEgressR
+		go func() {
+			sub, err := s.clientDeprecated.GetUpdateChannel(context.Background())
+			if err != nil {
+				logger.Errorw("failed to subscribe to results channel", err)
+			}
+
+			resChan := sub.Channel()
+			for {
+				select {
+				case msg := <-resChan:
+					b := sub.Payload(msg)
+					info := &livekit.EgressInfo{}
+					if err = proto.Unmarshal(b, info); err != nil {
+						logger.Errorw("failed to read results", err)
+						continue
+					}
+					s.handleUpdate(info)
+				case <-s.shutdown:
+					_ = sub.Close()
+					rs.Stop()
+					return
 				}
+			}
+		}()
+	}
 
-				switch res.Status {
-				case livekit.EgressStatus_EGRESS_COMPLETE,
-					livekit.EgressStatus_EGRESS_FAILED,
-					livekit.EgressStatus_EGRESS_ABORTED:
+	return nil
+}
 
-					// make sure endedAt is set so it eventually gets deleted
-					if res.EndedAt == 0 {
-						res.EndedAt = time.Now().UnixNano()
-					}
+func (s *EgressService) handleUpdate(info *livekit.EgressInfo) {
+	switch info.Status {
+	case livekit.EgressStatus_EGRESS_COMPLETE,
+		livekit.EgressStatus_EGRESS_FAILED,
+		livekit.EgressStatus_EGRESS_ABORTED:
 
-					err = s.es.UpdateEgress(context.Background(), res)
-					if err != nil {
-						logger.Errorw("could not update egress", err)
-					}
+		// make sure endedAt is set so it eventually gets deleted
+		if info.EndedAt == 0 {
+			info.EndedAt = time.Now().UnixNano()
+		}
 
-					// log results
-					if res.Error != "" {
-						logger.Errorw("egress failed", errors.New(res.Error), "egressID", res.EgressId)
-					} else {
-						logger.Infow("egress ended", "egressID", res.EgressId)
-					}
+		if err := s.es.UpdateEgress(context.Background(), info); err != nil {
+			logger.Errorw("could not update egress", err)
+		}
 
-					s.telemetry.EgressEnded(context.Background(), res)
+		// log results
+		if info.Error != "" {
+			logger.Errorw("egress failed", errors.New(info.Error), "egressID", info.EgressId)
+		} else {
+			logger.Infow("egress ended", "egressID", info.EgressId)
+		}
 
-				default:
-					err = s.es.UpdateEgress(context.Background(), res)
-					if err != nil {
-						logger.Errorw("could not update egress", err)
-					}
-				}
+		s.telemetry.EgressEnded(context.Background(), info)
 
-			case <-s.shutdown:
-				_ = sub.Close()
-				rs.Stop()
-				return
-			}
+	default:
+		if err := s.es.UpdateEgress(context.Background(), info); err != nil {
+			logger.Errorw("could not update egress", err)
 		}
+	}
+}
+
+// TODO: remove in future version
+func (s *EgressService) getFirst(f0, f1 func() (*livekit.EgressInfo, error)) (*livekit.EgressInfo, error) {
+	if s.clientDeprecated == nil {
+		return f1()
+	}
+
+	var infoV0, infoV1 *livekit.EgressInfo
+	var errV0, errV1 error
+	v0Done := make(chan struct{})
+	v1Done := make(chan struct{})
+
+	go func() {
+		infoV0, errV0 = f0()
+		close(v0Done)
 	}()
 
-	return nil
+	go func() {
+		infoV1, errV1 = f1()
+		close(v1Done)
+	}()
+
+	select {
+	case <-v0Done:
+		return infoV0, errV0
```

Comment:
I think there's a goroutine race here.. but when infoV0/errV0 are set without lock, the current thread may not be updated.

would recommend locking to be sure.

Reply:
switched to sending over channels